Firestore/core/src/firebase/firestore/remote/stream.mm

```diff
@@ -0,0 +1,318 @@
+                return;
+              }
+              strong_this->ResumeStartAfterAuth(maybe_token);
+            });
+      });
+}
+
+void Stream::ResumeStartAfterAuth(const StatusOr<Token>& maybe_token) {
+  EnsureOnQueue();
+
+  HARD_ASSERT(state_ == State::Starting,
+              "State should still be 'Starting' (was %s)", state_);
+
+  if (!maybe_token.ok()) {
+    OnStreamError(maybe_token.status());
+    return;
+  }
+
+  absl::string_view token = [&] {
+    auto token = maybe_token.ValueOrDie();
+    return token.user().is_authenticated() ? token.token()
+                                           : absl::string_view{};
+  }();
+
+  grpc_stream_ = CreateGrpcStream(datastore_, token);
+  grpc_stream_->Start();
+}
+
+void Stream::OnStreamStart() {
+  EnsureOnQueue();
+
+  state_ = State::Open;
+  DoOnStreamStart();
+}
+
+// Backoff
+
+void Stream::BackoffAndTryRestarting() {
+  EnsureOnQueue();
+
+  LOG_DEBUG("%s backoff", GetDebugDescription());
+
+  HARD_ASSERT(state_ == State::Error,
+              "Should only perform backoff in an error case");
+
+  state_ = State::Backoff;
+  backoff_.BackoffAndRun([this] { ResumeStartFromBackoff(); });
+}
+
+void Stream::ResumeStartFromBackoff() {
+  EnsureOnQueue();
+
+  HARD_ASSERT(state_ == State::Backoff, "Backoff elapsed but state is now: %s",
+              state_);
+
+  state_ = State::Initial;
+  Start();
+  HARD_ASSERT(IsStarted(), "Stream should have started.");
+}
+
+void Stream::InhibitBackoff() {
+  EnsureOnQueue();
+
+  HARD_ASSERT(!IsStarted(),
+              "Can only cancel backoff in a stopped state (was %s)", state_);
+
+  // Clear the error condition.
+  state_ = State::Initial;
+  ResetBackoff();
+}
+
+void Stream::ResetBackoff() {
+  EnsureOnQueue();
+  backoff_.Reset();
+}
+
+// Idleness
+
+void Stream::MarkIdle() {
+  EnsureOnQueue();
+
+  if (IsOpen() && !idleness_timer_) {
+    idleness_timer_ = worker_queue_->EnqueueAfterDelay(
+        kIdleTimeout, idle_timer_id_, [this] { Stop(); });
+  }
+}
+
+void Stream::CancelIdleCheck() {
+  EnsureOnQueue();
+  idleness_timer_.Cancel();
+}
+
+// Read/write
+
+void Stream::OnStreamRead(const grpc::ByteBuffer& message) {
+  EnsureOnQueue();
+
+  HARD_ASSERT(IsStarted(), "OnStreamRead called for a stopped stream.");
+
+  if (bridge::IsLoggingEnabled()) {
+    LOG_DEBUG("%s headers (whitelisted): %s", GetDebugDescription(),
+              Datastore::GetWhitelistedHeadersAsString(
+                  grpc_stream_->GetResponseHeaders()));
+  }
+
+  Status read_status = DoOnStreamRead(message);
+  if (!read_status.ok()) {
+    grpc_stream_->Finish();
+    // Don't expect GRPC to produce status -- since the error happened on the
+    // client, we have all the information we need.
+    OnStreamError(read_status);
+    return;
+  }
+}
+
+// Stopping
+
+void Stream::Stop() {
+  EnsureOnQueue();
+
+  LOG_DEBUG("%s stop", GetDebugDescription());
+
+  if (!IsStarted()) {
+    return;
+  }
+
+  RaiseGeneration();
```

Comment:
This is a different ordering than the other platforms and seems to be missing some steps. Notably, here we should be:
  * canceling the idle check
  * canceling the backoff timer
  * bumping the close count

This seems to only be doing the last. At a high level, we want to complete all of the bookkeeping for the state transition, including guaranteeing that timers aren't going to fire before making any callbacks.